cgen error when trying to use an embedded struct value, defined in all variant types of a sumtype

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 e4d24e8.844d89f
OS: linux, Ubuntu 20.04.6 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i3-3225 CPU @ 3.30GHz

getwd: /home/delian/code/v
vexe: /home/delian/code/v/v
vexe mtime: 2024-11-27 10:03:18

vroot: OK, value: /home/delian/code/v
VMODULES: OK, value: /home/delian/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.0
Git vroot status: weekly.2024.48-8-g844d89fd
.git/config present: true

CC version: cc (Ubuntu 10.5.0-1ubuntu1~20.04) 10.5.0
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg /home/delian/code/misc/2024_11_27__12/p.v`

```v
struct Value  { x int }
struct BValue { v ?Value }
struct Word     { BValue }
struct Long     { BValue }
struct Variadic { BValue }
type Param = Word | Long | Variadic
a := [ Param(Word{}) Long{}, Variadic{} ]
dump(a)
f := a[0]
v := f.v
dump(v)
```

**What did you expect to see?**

a compiled program

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/p.01JDPH6PATX4KMC7XAC6WCC9K7.tmp.c:1046: error: field 'v' has incomplete type
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzQ2ZWVhOWQyZWY0Y2JjYzQ2YmJkODIiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.qhXjI4FTVaiQdWL2EP0C-NfX_uOFZiwOtzSGCXrSQjA">Huly&reg;: <b>V_0.6-21425</b></a></sub>

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6937,7 +6937,7 @@ fn (mut g Gen) write_types(symbols []&ast.TypeSymbol) {
 				if sym.info.fields.len > 0 {
 					g.writeln('\t// pointers to common sumtype fields')
 					for field in sym.info.fields {
-						g.type_definitions.writeln('\t${g.styp(field.typ.ref())} ${c_name(field.name)};')
+						g.type_definitions.writeln('\t${g.styp(field.typ)}* ${c_name(field.name)};')
 					}
 				}
 				g.type_definitions.writeln('};')

--- a/vlib/v/tests/sumtypes/sumtype_with_embedded_struct_of_option_field_test.v
+++ b/vlib/v/tests/sumtypes/sumtype_with_embedded_struct_of_option_field_test.v
@@ -1,0 +1,30 @@
+struct Value {
+	x int
+}
+
+struct BValue {
+	v ?Value
+}
+
+struct Word {
+	BValue
+}
+
+struct Long {
+	BValue
+}
+
+struct Variadic {
+	BValue
+}
+
+type Param = Word | Long | Variadic
+
+fn test_sumtype_with_embedded_struct_of_option_field() {
+	a := [Param(Word{}), Long{}, Variadic{}]
+	dump(a)
+	f := a[0]
+	v := f.v
+	dump(v)
+	assert v == none
+}

